core/drivers/pcap.cc

```diff
@@ -1,193 +1,48 @@
 #include "pcap.h"
-struct snobj *PCAPPort::Init(struct snobj *conf) {
-  char errbuf[PCAP_ERRBUF_SIZE];
-  char *_dev = snobj_eval_str(conf, "dev");
-
-  if (!_dev) {
-    return snobj_err(EINVAL, "PCAP need to set dev option");
-  }
-
-  const std::string dev(_dev);
-
-  // non-blocking pcap
-  pcap_handle_ = pcap_open_live(dev.c_str(), PCAP_SNAPLEN, 1, -1, errbuf);
-  if (pcap_handle_ == nullptr) {
-    return snobj_err(ENODEV, "PCAP Open dev error: %s", errbuf);
-  }
 
-  int ret = pcap_setnonblock(pcap_handle_, 1, errbuf);
-  if (ret != 0) {
-    return snobj_err(ENODEV, "PCAP set to nonblock error: %s", errbuf);
+PCAPPort::~PCAPPort() {
+  if (pcap_handle_.IsInitialized()) {
+    DeInit();
   }
+}
 
-  LOG(INFO) << "PCAP: open dev " << dev;
-
+// don't use
+struct snobj *PCAPPort::Init(struct snobj *conf) {
   return nullptr;
 }
 
 pb_error_t PCAPPort::Init(const bess::protobuf::PCAPPortArg &arg) {
-  char errbuf[PCAP_ERRBUF_SIZE];
-
-  const std::string dev = arg.dev();
-  if (dev.length() == 0) {
-    return pb_error(EINVAL, "PCAP need to set dev option");
-  }
-  // Non-blocking pcap.
-  pcap_handle_ = pcap_open_live(dev.c_str(), PCAP_SNAPLEN, 1, -1, errbuf);
-  if (pcap_handle_ == nullptr) {
-    return pb_error(ENODEV, "PCAP Open dev error: %s", errbuf);
-  }
-
-  int ret = pcap_setnonblock(pcap_handle_, 1, errbuf);
-  if (ret != 0) {
-    return pb_error(ENODEV, "PCAP set to nonblock error: %s", errbuf);
+  if (pcap_handle_.IsInitialized()) {
+    return pb_error(EINVAL, "Device already initialized.");
   }
 
-  LOG(INFO) << "PCAP: open dev " << dev;
+  const std::string dev = arg.dev();
+  pb_error_t err;
+  pcap_handle_ = PCAPHandle(dev, &err);
 
-  return pb_errno(0);
+  return err;
 }
 
 void PCAPPort::DeInit() {
-  if (pcap_handle_) {
-    pcap_close(pcap_handle_);
-    pcap_handle_ = nullptr;
+  if (pcap_handle_.IsInitialized()) {
+    pcap_handle_.~PCAPHandle();
```

Comment:
You generally don't want to explicitly call a destructor like this.  Instead, you might make pcap_handle_ a pointer, call new above when constructing it, and then call delete here to destruct it.